validator/src/validator.rs

```diff
@@ -236,6 +243,55 @@ where
         self.init_block_producer(head_hash);
     }
 
+    /// Calculates the micro block producer timeout by averaging block times over a window ending at
+    /// `head_block`. The window size is dynamically adjusted if there are not enough blocks
+    /// available, i.e. at the beginning of the chain or if not all blocks are present in the
+    /// database.
+    fn compute_micro_block_producer_timeout(
+        head_block: &Block,
+        blockchain: &Blockchain,
+    ) -> Duration {
+        let end_block_number = head_block.block_number();
+        let end_timestamp = head_block.timestamp();
+
+        // Calculate the block number at the start of the window. Make sure this points to at least
+        // one block after the genesis block, as the genesis timestamp can be unreliable.
+        let start_block_number =
+            end_block_number.saturating_sub(Self::PRODUCER_TIMEOUT_WINDOW_SIZE);
+        let mut start_block_number =
+            u32::max(start_block_number, Policy::genesis_block_number() + 1);
+
+        // We might not have the block at `start_block_number` in our database. If it's missing,
+        // move up the chain to the next succeeding macro block(s), adjusting `start_block_number`
+        // in the process.
+        let start_timestamp = loop {
+            if let Ok(block) = blockchain.get_block_at(start_block_number, false, None) {
+                break block.timestamp();
+            }
+
+            // Try the next macro block.
+            start_block_number = Policy::macro_block_after(start_block_number);
+
+            // Bail if we didn't find a macro block. This shouldn't happen as we should always have
+            // at least the macro block at the beginning of our current batch.
+            if start_block_number >= end_block_number {
+                return Self::MIN_PRODUCER_TIMEOUT;
+            }
+        };
+
+        // Calculate the effective window size.
+        // If there are no blocks in the window, we return the default value.
+        let effective_window_size = end_block_number.saturating_sub(start_block_number) as u64;
+        if effective_window_size < 1 {
+            return Self::MIN_PRODUCER_TIMEOUT;
+        }
+
+        // Calculate the timeout and clamp it to the allowed range.
+        let avg_block_time = end_timestamp.saturating_sub(start_timestamp) / effective_window_size;
+        let timeout = Duration::from_millis(avg_block_time * 4);
```

Comment:
Magic Number.
I suggest putting it  as a const like the rest of them.

Reply:
Done